src/mpid/ch4/netmod/ofi/ofi_am_impl.c

```diff
@@ -8,75 +8,80 @@
 #include "ofi_noinline.h"
 
-int MPIDI_OFI_deferred_am_isend_issue(MPIDI_OFI_deferred_am_isend_req_t * dreq)
+static int issue_deferred_am_isend_eager(MPIDI_OFI_deferred_am_isend_req_t * dreq);
+
+static int issue_deferred_am_isend_eager(MPIDI_OFI_deferred_am_isend_req_t * dreq)
 {
     int mpi_errno = MPI_SUCCESS;
-    int dt_contig = 0;
-    MPI_Aint data_sz;
-    MPI_Aint dt_true_lb, last;
-    MPIR_Datatype *dt_ptr;
     char *send_buf;
-    bool need_packing = false;
-    bool do_eager = false;
-
-    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_OFI_DEFERRED_AM_ISEND_ISSUE);
-    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_DEFERRED_AM_ISEND_ISSUE);
-
-    MPIDI_Datatype_get_info(dreq->count, dreq->datatype, dt_contig, data_sz, dt_ptr, dt_true_lb);
-    need_packing = dt_contig ? false : true;
-    send_buf = (char *) dreq->buf + dt_true_lb;
+    MPI_Aint packed_size;
+    MPI_Aint dt_true_lb;
 
-    do_eager =
-        (MPIDI_OFI_AMREQUEST_HDR(dreq->sreq, am_hdr_sz) + dreq->data_sz
-         + sizeof(MPIDI_OFI_am_header_t) <= MPIDI_OFI_DEFAULT_SHORT_SEND_SIZE);
-
-    MPL_pointer_attr_t attr;
-    MPIR_GPU_query_pointer_attr(dreq->buf, &attr);
-    if (attr.type == MPL_GPU_POINTER_DEV && !MPIDI_OFI_ENABLE_HMEM) {
-        /* Force packing of GPU buffer in host memory */
-        need_packing = true;
-    }
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_OFI_ISSUE_DEFERRED_AM_ISEND_EAGER);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_ISSUE_DEFERRED_AM_ISEND_EAGER);
 
-    if (need_packing) {
+    if (dreq->need_packing) {
         /* FIXME: currently we always do packing, also for high density types. However,
          * we should not do packing unless needed. Also, for large low-density types
          * we should not allocate the entire buffer and do the packing at once. */
-        /* TODO: (1) Skip packing for high-density datatypes;
-         *       (2) Pipeline allocation for low-density datatypes; */
-        if (do_eager) {
-            mpi_errno = MPIDU_genq_private_pool_alloc_cell(MPIDI_OFI_global.pack_buf_pool,
-                                                           (void **) &send_buf);
-            if (send_buf == NULL) {
-                /* no buffer available, suppress error and exit */
-                mpi_errno = MPI_SUCCESS;
-                goto fn_exit;
-            }
-            MPIR_ERR_CHECK(mpi_errno);
-        } else {
-            /* only for large message RDMA read */
-            MPL_gpu_malloc_host((void **) &send_buf, data_sz);
+        /* TODO: (1) Skip packing for high-density datatypes; */
+        MPIR_Assert(dreq->data_sz <= MPIDI_OFI_DEFAULT_SHORT_SEND_SIZE);
+        mpi_errno = MPIDU_genq_private_pool_alloc_cell(MPIDI_OFI_global.pack_buf_pool,
+                                                       (void **) &send_buf);
+        if (send_buf == NULL) {
+            /* no buffer available, suppress error and exit */
+            mpi_errno = MPI_SUCCESS;
+            goto fn_exit;
```

Comment:
Shouldn't we check this first before anything else?

Reply:
I am not sure what you mean here? This check is already very early.

Reply:
The suppression of `mpi_errno` is removed. Just commenting here since I cannot find the previous discussion.

Reply:
The error suppression is still there for the case where genq does not return a buffer.

Reply:
Looks good now. I think a separate PR to address the error behavior of GenQ is probably a good idea.